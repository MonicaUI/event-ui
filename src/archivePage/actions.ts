import v1PageMigrations from 'nucleus-widgets/migrations/website/v1/pages';

export const LOAD_EVENT_ARCHIVE_PAGE_DATA = 'event-guestside-site/LOAD_EVENT_ARCHIVE_PAGE_DATA';

export const transformAndLoadEventArchivedPageData = (eventArchivePageData: $TSFixMe) => {
  return (dispatch: $TSFixMe, getState: $TSFixMe): $TSFixMe => {
    const { website: existingWebsite } = getState();
    dispatch({
      type: LOAD_EVENT_ARCHIVE_PAGE_DATA,
      payload: {
        website: {
          ...existingWebsite,
          pages: v1PageMigrations(eventArchivePageData.pages),
          layoutItems: eventArchivePageData.layoutItems,
          pluginData: eventArchivePageData.plugins
        }
      }
    });
  };
};
