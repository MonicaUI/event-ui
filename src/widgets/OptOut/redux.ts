/* eslint-disable import/no-cycle */

import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Opt-out and opt-in Redux module:
 * Event invitees can click "Opt Out" link in their emails to opt them out all communications from the event account
 */
const GET_OPT_OUT_STATUS = 'event-guestside-site/optOut/GET_OPT_OUT_STATUS';
const OPT_OUT_CONTACT = 'event-guestside-site/optOut/OPT_OUT_CONTACT';

const createGetOptedOutStatusAction = contactInfo => {
  return {
    type: GET_OPT_OUT_STATUS,
    payload: contactInfo
  };
};

export const getOptOutStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    const optedOutContact = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, guestId, contactId);
    dispatch(createGetOptedOutStatusAction(optedOutContact));
  };

export const createOptedOutAction = (contactInfo: $TSFixMe): $TSFixMe => {
  return {
    type: OPT_OUT_CONTACT,
    payload: contactInfo
  };
};

export const optOutAction =
  (optOutStatus: $TSFixMe, fromConfirmPage: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    // eslint-disable-next-line import/no-cycle
    const page = (await import(/* webpackChunkName optOut */ './content')).optInOutPageNew;
    if ((fromConfirmPage && optOutStatus) || !fromConfirmPage) {
      const optedOutContact = await eventGuestClient.optOut(
        accessToken,
        eventId,
        inviteeId,
        guestId,
        contactId,
        optOutStatus
      );
      dispatch(createOptedOutAction(optedOutContact));
    }
    dispatch(injectPageSnapshot(page));
  };

export const prepareForOptOutPageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    let page;
    if (!inviteeId && !guestId && !contactId) {
      page = (await import(/* webpackChunkName optOut */ './invalidInviteeContent')).invalidInviteeOptOut;
    } else {
      await dispatch(getOptOutStatus());
      if (getState() && getState().optOut && getState().optOut.optOutStatus) {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).optInOutPage;
      } else {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).confirmationPage;
      }
    }
    dispatch(injectPageSnapshot(page));
  };

export const optOutReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_OPT_OUT_STATUS:
    case OPT_OUT_CONTACT:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
