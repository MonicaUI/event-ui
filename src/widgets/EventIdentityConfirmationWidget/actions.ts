import {
  getEventRegistration,
  getVisibleRegistrationTypesByPlannerSelection
} from '../../redux/selectors/currentRegistrant';
import { getRegistrationPageFields } from '../../redux/selectors/currentRegistrationPath';
import { NOT_VISIBLE, READ_ONLY } from 'cvent-question-widgets/lib/DisplayType';
import { isWidgetReviewed } from '../../redux/website/pageContentsWithGraphQL';
import StandardContactFields from 'event-widgets/lib/StandardContactFields/StandardContactFields';
import { defaultRegistrationTypeId } from 'event-widgets/utils/registrationType';
import { CLEAR_REG_CART_INFERRED_FIELDS } from '../../redux/registrationForm/regCart/actionTypes';
import { setIn } from 'icepick';
import { getRegCart } from '../../redux/selectors/shared';
import { saveRegistration } from '../../redux/registrationForm/regCart';
import { runRegistrationTypeChangeValidationsForPrimaryAndGuest } from '../../redux/registrationForm/regCart';
import {
  handleRegistrationTypeSelectionConflict,
  validateUserRegistrationTypeSelection
} from '../../dialogs/selectionConflictDialogs';
import { withLoading } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import { getUpdateErrors } from '../../redux/registrationForm/errors';
import { openNoAdmissionItemAvailableForRegistrationTypeDialog, openPrivateEventErrorDialog } from '../../dialogs';
import { merge } from 'lodash';
import { populateVisibleProducts } from '../../redux/visibleProducts';
import { clearOutOfSyncUserSessionFieldsAndUrlQuery } from '../../redux/registrantLogin/actions';
import { ApolloClient } from '@apollo/client';
import type { RootState, AsyncAppThunk } from '../../redux/reducer';

export const isRegTypeReviewed = async (
  state: RootState,
  registrationTypeId: string,
  registrationPathId: string,
  apolloClient: ApolloClient<unknown>
): Promise<boolean> => {
  const widgetTypeName = 'RegistrationType';
  const visibleRegTypes = getVisibleRegistrationTypesByPlannerSelection(state, registrationPathId) || [];
  const isRegTypeWidgetReviewed = await isWidgetReviewed(state, { widgetType: widgetTypeName }, apolloClient);
  return visibleRegTypes.includes(registrationTypeId) && isRegTypeWidgetReviewed;
};

export function clearFields(): AsyncAppThunk {
  return async (
    dispatch,
    getState,
    {
      apolloClient
    }: {
      apolloClient?: ApolloClient<unknown>;
    }
  ) => {
    try {
      const state = getState();
      const eventRegistration = JSON.parse(JSON.stringify(getEventRegistration(state)));
      const isRegTypeWidgetReviewed = await isRegTypeReviewed(
        state,
        eventRegistration.registrationTypeId,
        eventRegistration.registrationPathId,
        apolloClient
      );
      if (state.userSession.regTypeId && !isRegTypeWidgetReviewed) {
        const regTypeChangeValidations = await runRegistrationTypeChangeValidationsForPrimaryAndGuest(
          validateUserRegistrationTypeSelection,
          defaultRegistrationTypeId,
          getState(),
          true,
          apolloClient
        );
        // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
        if (regTypeChangeValidations && regTypeChangeValidations.hasConflicts) {
          return dispatch(
            handleRegistrationTypeSelectionConflict(
              regTypeChangeValidations.validationResults,
              null,
              null,
              dissociateInviteeAndClearInferredFields
            )
          );
        }
      }
      const linkedInviteeIdToRemove = eventRegistration.attendee.attendeeId;
      if (state.userSession.regTypeId && !isRegTypeWidgetReviewed) {
        eventRegistration.registrationTypeId = defaultRegistrationTypeId;
      }
      /*
       * This is an existing pattern that we will need to revisit later. We should clear
       * the individual field rather than replace the entire reg cart
       */
      dispatch({
        type: CLEAR_REG_CART_INFERRED_FIELDS,
        payload: {
          regCart: setIn(
            getRegCart(state),
            ['eventRegistrations', eventRegistration.eventRegistrationId],
            eventRegistration
          )
        }
      });
      /**
       * on clear fields regType is cleared, need to reevaluate visible products
       */
      await dispatch(populateVisibleProducts(eventRegistration.eventRegistrationId));
      await dispatch(dissociateInviteeAndClearInferredFields());
      return await dispatch(withLoading(saveRegistration)(linkedInviteeIdToRemove));
    } catch (error) {
      if (
        // @ts-expect-error ts-migrate(2554) FIXME: Expected 1 arguments, but got 2.
        getUpdateErrors.isPrivateEvent(error, getState()) ||
        // @ts-expect-error ts-migrate(2554) FIXME: Expected 0-1 arguments, but got 2.
        getUpdateErrors.isAttendeeNotAllowedByCustomLogic(error, getState())
      ) {
        return await dispatch(openPrivateEventErrorDialog());
      }
      if (getUpdateErrors.isAdmissionItemsNotAvailableForRegTypeError(error)) {
        return await dispatch(openNoAdmissionItemAvailableForRegistrationTypeDialog());
      }
      throw error;
    }
  };
}

export function dissociateInviteeAndClearInferredFields(): AsyncAppThunk {
  return async (dispatch, getState) => {
    const state = getState();
    const eventRegistration = JSON.parse(JSON.stringify(getEventRegistration(state)));
    // invitee associated attendeeId and contact Id
    delete eventRegistration.attendee.attendeeId;
    delete eventRegistration.attendee.personalInformation.contactId;
    // inferred information to clear: 2. inviteeId related personalInformation
    const existingContactInfo = eventRegistration.attendee.personalInformation;
    const contactInfo = await Promise.all(
      getRegistrationPageFields(state).map(async field => {
        const isFieldReviewed = await isWidgetReviewed(state, { fieldId: field.fieldId });
        return field.display !== NOT_VISIBLE && field.display !== READ_ONLY && isFieldReviewed ? field : null;
      })
    );
    const filteredContactInfo = contactInfo.filter(Boolean);
    const copiedContactInfo = filteredContactInfo.map(field =>
      field.isCustomField
        ? {
            customFields: {
              [field.fieldId]: existingContactInfo.customFields[field.fieldId]
            }
          }
        : {
            // @ts-expect-error ts-migrate(2464) FIXME: A computed property name must be of type 'string',... Remove this comment to see the full error message
            [(StandardContactFields[field.fieldId] || {}).regApiPath]:
              // @ts-expect-error ts-migrate(2538) FIXME: Type 'string[]' cannot be used as an index type.
              existingContactInfo[(StandardContactFields[field.fieldId] || {}).regApiPath]
          }
    );
    const temp = copiedContactInfo.reduce(
      (copiedInfo, info) => (info.customFields ? merge(copiedInfo, info) : Object.assign(copiedInfo, info)),
      {}
    );
    eventRegistration.attendee.personalInformation = temp;

    await dispatch(clearOutOfSyncUserSessionFieldsAndUrlQuery());
    // Reversing the order so that inviteeId is set to undefined for sourceId to be shown as editable.
    dispatch({
      type: CLEAR_REG_CART_INFERRED_FIELDS,
      payload: {
        regCart: setIn(
          getRegCart(state),
          ['eventRegistrations', eventRegistration.eventRegistrationId],
          eventRegistration
        )
      }
    });
  };
}
