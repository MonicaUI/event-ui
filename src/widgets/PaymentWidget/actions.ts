/* Method for adding / removing taxes in Order Summary for planner reg */
import { getRegCart } from '../../redux/selectors/shared';
import {
  DISCOUNT_CODE_VALIDATION_FAILURE,
  UPDATE_REG_CART_FAILURE,
  UPDATE_REG_CART_PENDING,
  UPDATE_REG_CART_SUCCESS
} from '../../redux/registrationForm/regCart/actionTypes';
import { closeDialogContainer, hideLoadingOnError } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import { unset, updateIn } from 'icepick';
import {
  getAllowableEventRegistrationsForDiscountCapacityMap,
  getAvailableCapacity,
  getRegistrantsCountedTowardsDiscountCapacity,
  getRegistrantsEligibleForDiscount,
  getRegistrationTypeId,
  getRegistrationPathId
} from '../../redux/registrationForm/regCart/selectors';
import { findKnownErrorResourceKey, getUpdateErrors } from '../../redux/registrationForm/errors';
import { loadAvailableCapacityCounts } from '../../redux/capacity';
import { getEventRegistrationId, isGroupRegistration } from '../../redux/selectors/currentRegistrant';
import { openGuestProductSelectionDialog, openAlreadyRegisteredDialog } from '../../dialogs';
import Logger from '@cvent/nucleus-logging';
import { updateRegCart } from '../../redux/registrationForm/regCart';

const LOG = new Logger('widgets/PaymentWidget/actions');

export function updateIgnoreTaxesInRegCart(ignoreTaxes: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      accessToken,
      clients: { regCartClient },
      event,
      account
    } = state;
    const regCart = getRegCart(state);
    let response;
    dispatch({ type: UPDATE_REG_CART_PENDING });
    try {
      response = await updateRegCart(state, regCartClient, accessToken, {
        ...regCart,
        ignoreTaxes
      });
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });
      LOG.debug(`updateRegCart with id ${regCart.regCartId} was successful`);
    } catch (error) {
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(regCart, getEventRegistrationId(getState())),
          getRegistrationPathId(regCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      throw error;
    }
  };
}

export function updateIgnoreServiceFeesInRegCart(ignoreServiceFee: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      accessToken,
      clients: { regCartClient },
      event,
      account
    } = state;
    const regCart = getRegCart(state);
    let response;
    dispatch({ type: UPDATE_REG_CART_PENDING });
    try {
      response = await updateRegCart(state, regCartClient, accessToken, {
        ...regCart,
        ignoreServiceFee
      });
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });
      LOG.debug(`updateRegCart with id ${regCart.regCartId} was successful`);
    } catch (error) {
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(regCart, getEventRegistrationId(getState())),
          getRegistrationPathId(regCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      throw error;
    }
  };
}

/**
 * applies a discount code to the reg cart
 */
export function updateDiscountCodes(
  discountCode: $TSFixMe,
  removeDiscount: $TSFixMe,
  eventRegistrationsToApplyCodeTo?: $TSFixMe,
  previousAvailableCapacity?: $TSFixMe,
  previousErrors?: $TSFixMe
) {
  // eslint-disable-next-line complexity
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      regCartStatus: { lastSavedRegCart },
      clients: { regCartClient },
      event,
      account
    } = getState();
    dispatch({ type: UPDATE_REG_CART_PENDING });
    const regCart = getRegCart(getState());

    /*
     * when we try to apply new discount we need to use the latest reg cart, so we can calculate the
     * discount based on current reg cart info, this is needed for discount advance filter
     */
    let updatedRegCart = {};
    if (removeDiscount) {
      updatedRegCart = updateIn(lastSavedRegCart, ['discounts'], discountCodeMap =>
        unset(discountCodeMap, discountCode.toLowerCase())
      );
    } else if (eventRegistrationsToApplyCodeTo) {
      updatedRegCart = updateIn(regCart, ['discounts'], discountCodeMap => {
        return { ...discountCodeMap, [discountCode]: { discountCode, eventRegistrationsToApplyCodeTo } };
      });
    } else {
      updatedRegCart = updateIn(regCart, ['discounts'], discountCodeMap => {
        return { ...discountCodeMap, [discountCode]: { discountCode } };
      });
    }
    try {
      const response = await updateRegCart(getState(), regCartClient, accessToken, updatedRegCart);
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });

      LOG.debug(`updateRegCart with id ${lastSavedRegCart.regCartId} was successful`);
    } catch (error) {
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(updatedRegCart, getEventRegistrationId(getState())),
          getRegistrationPathId(updatedRegCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      let validationError = null;

      const state = getState();
      // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
      const errors = state.registrationForm && state.registrationForm.errors;
      const currentAvailableCapacity = getAvailableCapacity(errors);

      if (getUpdateErrors.isDiscountCapacityInsufficient(error)) {
        if (previousAvailableCapacity && previousAvailableCapacity !== currentAvailableCapacity) {
          await dispatch(selectDiscountForRegistrant(discountCode, previousErrors));
        } else {
          await dispatch(selectDiscountForRegistrant(discountCode, null));
        }
        return;
      } else if (getUpdateErrors.isProductAvailabilityError(error)) {
        // TODO: display error specific for discount capacity
        dispatch(loadAvailableCapacityCounts());
        validationError = 'EventWidgets_Validations_DiscountCodeCapacityFull__resx';
      } else if (getUpdateErrors.isKnownError(error)) {
        validationError = findKnownErrorResourceKey(error.responseBody.validationMessages);
      }

      if (validationError) {
        return dispatch({
          type: DISCOUNT_CODE_VALIDATION_FAILURE,
          payload: {
            validationError
          }
        });
      }

      throw error;
    }
  };
}

export function updateRegCartForAutoDiscounts() {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      clients: { regCartClient },
      event,
      account
    } = getState();
    const regCart = getRegCart(getState());
    dispatch({ type: UPDATE_REG_CART_PENDING });
    try {
      const response = await updateRegCart(getState(), regCartClient, accessToken, regCart);
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });
      LOG.debug(`updateRegCart with id ${regCart.regCartId} was successful`);
    } catch (error) {
      if (getUpdateErrors.isInviteeAlreadyRegistered(error)) {
        LOG.debug(`updateRegCart with id ${regCart.regCartId} failed`, error);
        dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
        dispatch(hideLoadingOnError());
        return dispatch(
          openAlreadyRegisteredDialog({
            title: 'EventGuestSide_ExistingInvitee_AlreadyRegistered_Header__resx',
            instructionalText: 'EventGuestSide_ExistingInvitee_AlreadyRegistered_InstructionalText__resx',
            prepopulateForm: true
          })
        );
      }
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(regCart, getEventRegistrationId(getState())),
          getRegistrationPathId(regCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      throw error;
    }
  };
}

/**
 * Apply discount to selected registrants only
 */
function selectDiscountForRegistrant(discountCodeValue, previousErrors) {
  return async (dispatch, getState) => {
    const {
      registrationForm,
      regCartStatus: { lastSavedRegCart }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const errors = previousErrors || (registrationForm && registrationForm.errors);
    const allowableEventRegistrationsForDiscountCapacityMap =
      getAllowableEventRegistrationsForDiscountCapacityMap(errors);
    // Get only those registrants who count towards discount capacity
    const discountCapacityRegistrants = getRegistrantsCountedTowardsDiscountCapacity(
      lastSavedRegCart,
      allowableEventRegistrationsForDiscountCapacityMap
    );
    dispatch(handleDiscountSelectionForInsufficientCapacity(discountCodeValue, discountCapacityRegistrants));
  };
}

/* Apply discount to registrants selected in the modal */
const applyDiscountToSelectedRegistrants = (
  discountCodeValue,
  currentPrimaryRegId,
  selectedEventRegIds,
  eventRegistrations
) => {
  return async (dispatch, getState) => {
    dispatch(closeDialogContainer());
    const {
      registrationForm,
      regCartStatus: { lastSavedRegCart }
    } = getState();
    const selectedDiscountEligibleRegistrants = [];
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const errors = registrationForm && registrationForm.errors;
    const allowableEventRegistrationsForDiscountCapacityMap =
      getAllowableEventRegistrationsForDiscountCapacityMap(errors);
    const availableCapacity = getAvailableCapacity(errors);
    // Get all registrants who are eligible for the discount
    let discountEligibleRegistrants = [];
    eventRegistrations.forEach(eventReg => {
      const eventRegId = eventReg.eventRegistrationId;
      if (selectedEventRegIds[eventRegId].isSelected) {
        // If this registrant is selected, apply discount code as applicable
        discountEligibleRegistrants = getRegistrantsEligibleForDiscount(
          lastSavedRegCart,
          allowableEventRegistrationsForDiscountCapacityMap,
          eventRegId
        );
        discountEligibleRegistrants.forEach(regId => {
          selectedDiscountEligibleRegistrants.push(regId);
        });
      }
    });
    dispatch(
      updateDiscountCodes(discountCodeValue, false, selectedDiscountEligibleRegistrants, availableCapacity, errors)
    );
  };
};

function handleDiscountSelectionForInsufficientCapacity(discountCodeValue, discountCapacityRegistrants) {
  return async (dispatch, getState) => {
    const state = getState();
    const eventRegistrations = discountCapacityRegistrants;
    /*
     * this variable is needed for the dialog logic, but since we don't need it for the discount logic,
     * we pass '' and check for it in the dialog logic
     */
    const eventRegistrationId = '';
    // Passing discountCodeValue into modal's productId field
    const discountId = discountCodeValue;
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const errors = state.registrationForm && state.registrationForm.errors;
    const availableCapacity = getAvailableCapacity(errors);
    const discountTitle = 'GuestProductSelection_DiscountTitle__resx';
    // overrideCapacity will always be false in this case
    const overrideCapacity = false;
    const isGroupReg = isGroupRegistration(state);
    const eventRegSelections = {};
    // the eventRegSelections properties would always be false here
    eventRegistrations.forEach(eventReg => {
      const eventRegId = eventReg.eventRegistrationId;
      eventRegSelections[eventRegId] = {
        isSelected: false,
        isDisabled: false,
        registeredForProductInGroup: false
      };
    });
    await dispatch(
      openGuestProductSelectionDialog(
        'GuestProductSelection_ApplyDiscount__resx',
        discountId,
        discountTitle,
        availableCapacity,
        overrideCapacity,
        eventRegSelections,
        eventRegistrations,
        eventRegistrationId,
        isGroupReg,
        applyDiscountToSelectedRegistrants
      )
    );
  };
}
