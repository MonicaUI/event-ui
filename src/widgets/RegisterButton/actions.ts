import { routeToPage, redirectToPage } from '../../redux/pathInfo';
import { logoutRegistrant } from '../../redux/registrantLogin/actions';
import { openStartNewRegistrationDialogDuringRegistration } from '../../dialogs';
import * as currentRegistrant from '../../redux/selectors/currentRegistrant';
import { DECLINE, GUEST_REGISTRATION, REGISTRATION } from '../../redux/website/registrationProcesses';

/**
 * @param props.config.text Registration dialog text
 * @param props.style Styling for registration dialog
 * @param props.classes Classnames for registration dialog
 * Based on user state and current page
 * Dispatch actions either to redirect to registration start,
 * logout and redirect to 'register' page, or prompt with new registration dialog.
 * RegisterNow widget can exist in each of the conditions, all actions are available
 * RegisterAnother widget can only exists on post-reg pages,
 * openStartNewRegistrationDialog is the only available action
 */
export const promptToStartNewRegistration = (props: $TSFixMe) => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const loggedIn = currentRegistrant.isLoggedIn(state);
    const registrationPage = REGISTRATION.isTypeOfCurrentPage(state) || GUEST_REGISTRATION.isTypeOfCurrentPage(state);
    const registrationDeclinePage = DECLINE.isTypeOfCurrentPage(state);
    const titleText = props.config.text.htmlContent
      ? props.config.text.htmlContent.replace(/<[^>]*>/g, '')
      : props.config.text;
    if (!loggedIn && !registrationPage && !registrationDeclinePage) {
      return dispatch(routeToPage('register'));
    }
    if (registrationDeclinePage) {
      await dispatch(logoutRegistrant());
      return dispatch(redirectToPage('register'));
    }
    return dispatch(
      openStartNewRegistrationDialogDuringRegistration({
        title: titleText,
        classes: { ...props.classes },
        style: props.style,
        startWithDefaultRegPath: props.startWithDefaultRegPath || false
      })
    );
  };
};
