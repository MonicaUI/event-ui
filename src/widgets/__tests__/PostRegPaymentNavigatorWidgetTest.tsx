import React from 'react';
import { loginRegistrant } from '../../redux/registrantLogin/actions';
jest.mock('../../redux/pathInfo');
jest.mock('../../redux/registrationForm/regCart', () => ({
  saveRegistration: x => x,
  finalizeCancelRegistration: x => x
}));
jest.mock('../../redux/registrantLogin/actions', () => ({
  logoutRegistrant: () => {},
  loginRegistrant: jest.fn()
}));
jest.mock('../../redux/website', () => {
  return {
    ...jest.requireActual<$TSFixMe>('../../redux/website'),
    __esModule: true,
    getDefaultWebsitePageId: () => 'website1'
  };
});

import { shallow } from 'enzyme';
import PostRegPaymentNavigatorWidgetTest from '../RegistrationPostRegPaymentNavigatorWidget';
import LinearPageNavigatorWidget from 'nucleus-widgets/lib/LinearPageNavigator/LinearPageNavigatorWidget';
import { REGISTERING } from '../../redux/registrationIntents';
import { createLocalNucleusForm } from 'nucleus-form/src/NucleusForm';
import pageContainingWidgetFixture from '../../testUtils/pageContainingWidgetFixture';
import * as InviteeStatus from 'event-widgets/utils/InviteeStatus';
import EventGuestClient from '../../clients/EventGuestClient';
import { mount } from 'enzyme/build/index';
import Form from 'nucleus-form/src/components/Form';
import { Grid } from 'nucleus-core/layout/flexbox';
import { Provider } from 'react-redux';
import { routeToPage } from '../../redux/pathInfo';

jest.mock('../../clients/EventGuestClient');
jest.mock('../../utils/confirmationUtil', () => ({
  getConfirmationPageIdForInvitee: () => () => Promise.resolve('confirmationPageId')
}));

(loginRegistrant as $TSFixMe).mockImplementation(() => {
  return dispatch1 => {
    dispatch1({
      type: '[MOCK]/loginRegistrant',
      payload: {}
    });
  };
});

const eventPersonaClient = {
  identifyInvitee: () => Promise.resolve({ inviteeStatus: InviteeStatus.NoResponse })
};

function getState() {
  return {
    text: { translate: x => x },
    registrationForm: {
      warnings: {},
      regCart: {
        eventRegistrations: {
          REG1: {
            id: 'REG1',
            attendeeType: 'ATTENDEE',
            attendee: {
              personalInformation: {
                emailAddress: 'tony@stark.com'
              }
            },
            confirmationNumber: 'CONFIRMNUM'
          }
        }
      }
    },
    regCartStatus: {
      registrationIntent: REGISTERING
    },
    defaultUserSession: {
      isPlanner: false
    },
    plannerRegSettings: {
      exitUrl: 'https://www.google.com/'
    },
    event: {
      registrationTypes: {
        '00000000-0000-0000-0000-000000000000': {
          id: '00000000-0000-0000-0000-000000000000'
        }
      }
    },
    website: {
      ...pageContainingWidgetFixture('cancellationPageId', 'widgetId'),
      pluginData: {
        registrationProcessNavigation: {
          registrationPaths: {
            regPathId: {
              id: 'regPathId',
              pageIds: ['pageId'],
              cancellationPageIds: ['cancellationPageId'],
              confirmationPageId: 'confirmationPageId'
            }
          }
        }
      }
    },
    appData: {
      registrationSettings: {
        registrationPaths: {
          regPathId: {
            id: 'regPathId',
            cancellation: {
              enabled: true
            },
            isDefault: true
          }
        }
      }
    },
    postRegistrationPaymentData: {
      isCheckingOut: false
    },
    clients: { eventPersonaClient, eventGuestClient: new EventGuestClient() }
  };
}
async function dispatch(action) {
  if (typeof action === 'function') {
    return await action(dispatch, getState);
  }
}
const subscribe = () => {};
const defaultProps = {
  nucleusForm: createLocalNucleusForm(),
  config: {
    displayText: {
      backward: 'backward',
      forward: 'forward',
      complete: 'complete',
      exit: 'exit'
    }
  },
  classes: {},
  style: {},
  translate: x => x,
  store: { dispatch, subscribe, getState },
  id: 'widgetId'
};

let mockStore;
const mountWidget = () => {
  mockStore = { dispatch, subscribe, getState };
  return mount(
    <Provider store={mockStore}>
      <Form>
        <Grid>
          <PostRegPaymentNavigatorWidgetTest {...defaultProps} />
        </Grid>
      </Form>
    </Provider>
  );
};

describe('Tests for RegistrationPostRegPaymentNavigatorWidget', () => {
  test('Checking if it is rendering properly', () => {
    const component = mountWidget();
    expect(component).toMatchSnapshot('Widget Renders');
  });
});

describe('RegistrationNavigatorWidget', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('navigates to the confirmation page on exit', async () => {
    const wrapper = shallow(<PostRegPaymentNavigatorWidgetTest {...defaultProps} />).dive();
    const widget = wrapper.find(LinearPageNavigatorWidget);
    await widget.props().onExitRequest();
    expect(routeToPage).toHaveBeenCalledWith('confirmationPageId');
  });
});
