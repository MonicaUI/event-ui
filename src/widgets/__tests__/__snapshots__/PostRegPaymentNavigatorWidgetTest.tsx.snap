// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for RegistrationPostRegPaymentNavigatorWidget Checking if it is rendering properly: Widget Renders 1`] = `
<div>
  <div
    class=""
  >
    <div />
  </div>
</div>
`;
